### DISCUSSION TITLE
miekg/dns parser

### DISCUSSION THREAD

**Discussion by gallypette** [dmachard/DNS-collector#1224] [General]
Hi,

I see that `go-dnscollector` already uses or mentions `miekg/dns` in some parts and I was wondering why it wasn't using its parser? This could prove useful as for instance `miekg/dns` support all `rdatatype` whereas here we only parse a few.

see https://github.com/miekg/dns/blob/master/scan_rr.go
see https://github.com/miekg/dns/blob/master/parse_test.go

**Comment by dmachard** [comment]
Hi,

I asked myself this question  at the beginning and I did the following choice (with my experience from Python language perhaps it was not a good idea, I accept that ) 

=> it's not a dns server so decodes just the minimum, not more!

I known it's not a good idea to do that from scratch but I learned a lot (and it was fun), and with the contribution of @jtt too.
Also I wanted to learn how to decode DNS packets , generate malformed packets and how to protect against them.

Finally, we are running it in production since the 2 years without issues regarding the DNS parser so it stayed like that.

For the future, it will be nice to introduce the miekg/dns parser to compare the memory footprint and cpu usage for example 
If the embedded DNS parser is bad regarding performance, I will be happy to switch to miekg/dns lib :)

Here the story.

Denis

**Comment by gallypette** [comment]
Understood.
We have a slightly different take because (1) we use this project for threat intel and some less-used rtypes have value in this context and (2) miekg/dns is widely used and fuzzed methodically so that we can have a reasonable guarantee that ingesting this data won't lead to smth nasty.
 
I will try to have a poc with the miekg to see how it impact performances.

**Comment by dmachard** [comment]
FYI, below a comparaison regarding performance 

| Parser | Speed (ns/op) | Memory (B/op) | Allocations (allocs/op) |
|--------|----------------|----------------|-------------------------|
| **Embedded DNS parser DNS-collector** | **~85 ns** | **128 B** | **4 allocs** |
| `miekg/dns` Parser | ~180 ns | 176 B | 7 allocs |

Note: Benchmarks run on AMD Ryzen 9 9900X CPU. The custom parser is ~2.1x faster and consumes less memory with fewer allocations.
